drizzlepac/hapsequencer.py

```diff
@@ -1122,49 +1146,76 @@ def update_active_wcs(filename, wcsname, keyword_wcs_names_dict, image_dict):
     # generate a combined WCS
     num_sci_ext, extname = util.count_sci_extensions(filename)
     extname_list = [(extname, x+1) for x in range(num_sci_ext)]
 
-    hdu = fits.open(filename, mode="update")
+    hdu = fits.open(filename, mode='update')
 
     # Check if the desired WCS solution is already the active solution
     # whereupon there is nothing to do
-    key = wcsutil.altwcs.getKeyFromName(hdu[1].header, wcsname)
-    if key != " ":
-
-        # If the current active solution is not already saved in the file, it
-        # must be stored as an alternate solution
-        wcs_list = image_dict[filename]
-        is_stored = [item for item in wcs_list if wcsname == item]
-        if not is_stored:
-            out_tuple = wcsutil.altwcs.archive_wcs(filename, extname_list, wcsname = wcsname)
-            log.info("Archiving previous active WCS solution: {}".format(out_tuple))
-
-        # Is the source of the wcsname for this image from keywords or a headerlet?
-        # The source dictates how the WCS will be made the active WCS
-        # Examine the "keyword" WCS solutions as a full string match...
-        keyword_wcs_list = keyword_wcs_names_dict[filename]
+    key = wcsutil.altwcs.getKeyFromName(hdu['SCI', 1].header, wcsname)
+
+    # Case where the desired active solution is not the current active
+    # solution
+    if key != ' ':
+        # If the current active WCS solution is not already archived, this method will do it
+        current_active_hdrname = hdu['SCI', 1].header['HDRNAME']
+        current_active_wcsname = hdu['SCI', 1].header['WCSNAME']
+        wcsutil.headerlet.archive_as_headerlet(filename, current_active_hdrname, wcskey='PRIMARY')
+        log.info("Archiving previous active WCS solution as necessary: {}".format(current_active_wcsname))
+
+        # Get the distortion model identification of the desired active WCS solution
+        index = wcsname.upper().find('IDC_')
+        idc_new_string = ''
+        if index > 0:
+            idc_new_string = wcsname[0:13]
+
+        # Examine the alternate WCS solutions to determine if they will be auto-archived due
+        # to a distortion model change
+        wcs_key_dict = wcsutil.altwcs.wcsnames(filename, ext=1)
+        for wkey, wname in wcs_key_dict.items():
+            index = wname.upper().find(idc_new_string.upper())
+
+            # No match so solution will be copied to a headerlet automatically when the new primary is set
+            if index == -1 and wkey.upper() != 'O':
+                log.info("Archiving alternate WCS solution as a headerlet as necessary: {}".format(wname))
+
+                # Now check if the HDRNAME between this solution and a headerlet already exists
+                # Get the headerlet HDRNAMES fresh every time
+                hdr_keyword = hdu[1].header['HDRNAME' + wkey.upper()]
+                headerlet_hdrnames = wcsutil.headerlet.get_headerlet_kw_names(filename, kw="HDRNAME")
```

Comment:
The set of headerlet_hdrnames could be determined prior to this loop and, I think, eliminate the need to do this again in line 1204, if I understand what is going on correctly...

Reply:
True. This is cruft.